### PR TITLE
RFC

### PR DESCRIPTION
**Pull request by lezcano**
I left a number of comments in bold throughout the text that occurred to me as I was writing this.

### CODE REVIEW

File: RFC.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+found. Iterating this process, we ended up with a small list of differences
+between the NumPy and the PyTorch API which we sorted out by hand and finished
+implementing. That list and the prioritization discussion can be found in
+[the first few posts of this issue](https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/87).
+
+The second point of preserving NumPy semantics as much as possible will be used
+in the sequel to discuss some points like the default dtypes that are used
+throughout the implementation.
+
+### Annotation-based preprocessing
```
Comment (rgommers) [inline RFC.md:127]: This part is mostly about _implementation_, while the next part (`### Dtypes`) is about _user-observable behavior_. Those two topics should probably be more explicitly separated, and switched around in the document. I'll also note that there's an RFC template that would be good to adhere to, and which means some refactoring of this doc: https://github.com/pytorch/rfcs/blob/master/RFC-0000-template.md

**Review by rgommers — COMMENTED** [review]
Thanks Mario! Here are my first set of high-level comments that seem worth discussing.

File: RFC.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+```python
+import torch
+x = torch.randn(3, 4, dtype=torch.float64)
+y = torch.randn(4, 3, dtype=torch.float64)
+z = torch.matmul(x, y)
+w = z.sum()
+```
+
+Here we already see a couple differences between NumPy and PyTorch. The most
```
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:49]: I pushed some changes. I think that leaving these in this example and moving the FKA `## DTypes` section right after this section makes these points here have more meaning.

This is how I had it at the beginning, which makes more sense, but then I moved the dtypes section down the document and left this part a bit on its own.

File: RFC.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+- `AxisLike`: Takes anything that can be accepted as an axis (e.g. a tuple or
+  an `ndarray`) and returns a tuple.
+- `OutArray`: Asserts that the input is a `torch_np.ndarray`. This is used
+  to implement the `out` arg.
+
+Note that none of the code in this implementation  makes use of NumPy. We are
+writing `torch_np.ndarray` above to make more explicit our intents, but there
+shouldn't be any ambiguity.
+
+**OBS(Lezcano)**: `DTypeLike` should be `Optional[DTypeLike]`
```
Comment (ev-br) [inline RFC.md:266]: `dtype=None` means a default dtype at least in numpy, so None is also a dtype_like as it were, no?
Can change it of course if desired.

File: RFC.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+Note that none of the code here makes use of NumPy. We are writing
+`torch.numpy.ndarray` above to make more explicit our intents, but there
+shouldn't be any ambiguity here.
+
+**OBS(Lezcano)**: `DTypeLike` should be `Optional[DTypeLike]`
+**OBS(Lezcano)**: Should we have a `NotImplementedType` to mark the args that
+are not being implemented? We could then assert that either that parameter has
+not been provided, and if it has, it has the same value as the default. The
+goal here would be to either use all the args of a function in its
+implementation, or mark explicitly those that we don't use.
```
[the remark this answers is not in the document]
Comment (ev-br) [reply RFC.md:271]: gh-114

**Review by ev-br — COMMENTED** [review]
A few minor comments on a first read-through.

File: RFC.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+```python
+import torch
+x = torch.randn(3, 4, dtype=torch.float64)
+y = torch.randn(4, 3, dtype=torch.float64)
+z = torch.matmul(x, y)
+w = z.sum()
+```
+
+Here, we can already spot a couple differences between NumPy and PyTorch.
```
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:47]: That's a good point. I'll add a point on randomness below.

File: RFC.md [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+Note that this code mixing NumPy and PyTorch already works, as `torch.Tensor`
+implements the `__array__` method. Now, the compatibility layer allows us to
+trace through it. In order to do that, there would be no necessary changes,
+other than simply ask `torch.compile` to trace through it:
+
+```python
+@compile
+def fn(x, y):
+    return np.multiply(x, y).sum()
+```
```
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:85]: As it's discussed at the beginning as "the only goal of this project", using the compat layer would allow to compile the code into a graph.

Once you have this graph, you can do many more things other than running it with CPU tensors. For once, you can compile it into CUDA and run it with CUDA tensors. Or even, on TPUs. You can also simply run it without compiling the graph and actually executing PyTorch ops one at a time as you'd do executing eager.

There is also a proposal to expose this API to the public via `torch.numpy`, so that people can play around more directly with it, but core maints had some mixed feelings about that. I wrote a few pros and cons in the Unresolved Questions below.

File: RFC.md [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+### Design decisions
+
+The two main ideas driving the design of this compatibility layer are the following:
+
+1. The behavior of the layer should be as close to that of NumPy as possible
+2. The layer follows NumPy master
```
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:92]: Most recent release seems better, yep.

File: RFC.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ev-br) [reply RFC.md:85]: My point here (poorly worded, yes) is only that on a reading this from a blank slate my knee-jerk reaction was along the lines of it just said that this works already, what's the need for anything else then. So it's only about the wording. I don't have an immediate suggestion on how to tweak it though.

File: RFC.md [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+they simply accept `torch.Tensor` and PyTorch-land objects and return
+`torch.Tensor`s. For example, we would implement `np.diag` as
+
+```python
+def diag(v, k=0):
+    return torch.diag(v, k)
+```
+
+In this layer, if a NumPy function is composite (calls other NumPy functions
+internally), we can simply vendor its implementation, and have it call our
```
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:232]: Yes. We didn't that too often, but it could be done if we wanted. Say someone finds a bug in a function, and they find that the NumPy function is written in Python. A pretty good patch would be to simply bring the NumPy implementation into the compat layer.

Or is there any licensing issue or otherwise with doing this?

File: RFC.md (hunk 3 above)
Comment (lezcano) [reply RFC.md:266]: I think the best way to think about these typing things is to think about them in a strongly typed language. For example, in C++, this would clearly be a `std::optional<DTypeLike>` with `enum class DTypeLike  { Int32, ...}`. The current option would be implemented by having a `DTypeLike::null`, which would not be pretty.

File: RFC.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:85]: Fair point. I'll modify it slightly to point out that, it does work for CPU tensors, but not for CUDA or on TorchDynamo.

File: RFC.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ev-br) [reply RFC.md:232]: <hairsplitting>This layer's content is composable irrespective of whether a particular implementation is vendored from numpy or not, so maybe stress it? Sure, we sometimes copy-paste from the numpy source tree, is it a big deal</hairsplitting>

File: RFC.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:85]: I'm actually going to leave it almost as is, because even though we say that the example currently works with tensors, we, right after, then say "Now, the compatibility layer allows us to trace through it".

File: RFC.md [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+NumPy defaults whenever the `dtype` was not made explicit in a factory function.
+
+**TODO(Lezcano)**: I just realized that we do not have a clean way to change
+the default dtype of `torch_np` to those from PyTorch. We should implement
+that utility flag, similar to
+[`torch.set_default_dtype`](https://pytorch.org/docs/stable/generated/torch.set_default_dtype.html).
+Perhaps call it `torch_np.use_torch_defaults()` and then add a way for users
+to be able to set their own int/float/complex defaults.
+**TODO(Lezcano)**: Do we just use them just in factory functions, or do we also
+use them anywhere else -> Check
```
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:111]: Yeah, I think this is good enough. Removing this comment.

[the remark this answers is not in the document]
Comment (ev-br) [reply RFC.md]: NB: I'm preparing a PR with a `set_default_dtype`.

File: RFC.md [hunk 10]
```diff
@@ -129,8 +130,8 @@ NumPy scalars as 0-D arrays. This may cause small divergences in some cases like
 array([2, 4, 6])
 ```
 
-but we don't expect these to pose a big issue in practice. Note that in this
-implementation `torch_np.int32(2)` would return the same as `torch_np.asarray(2)`.
+but we don't expect these to pose a big issue in practice. Note that in the
+proposed implementation `np.int32(2)` would return the same as `np.asarray(2)`.
```
Comment (ev-br) [inline RFC.md:132]: in the proposed implementation --- when compiled through TorchDynamo via `@torch.compile`, correct?

Otherwise it make a naive reader in me utterly confused.

File: RFC.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (lezcano) [reply RFC.md:240]: That's discussed in the next sentence:
>In other words, at this level, functions are composable, as they are simply regular PyTorch functions.

File: RFC.md [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+
+## Summary
+This RFC describes a proposal for a translation layer from NumPy into PyTorch.
+In simple terms, this accounts for implementing most of NumPy's API (`ndarray`,
+the `numpy`, `numpy.linalg`, `numpy.fft`  modules, etc) using `torch.Tensor`
+and PyTorch ops as backend.
+
+
+The this project has a main goal as per the
+[initial design document](https://docs.google.com/document/d/1gdUDgZNbumFORRcUaZUVw790CtNYweAM20C1fbWMNd8):
```
Comment (ev-br) [inline RFC.md:16]: The access is gated.

File: RFC.md [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+
+Then, we wrap these Python-land functions with a `normalizer` decorator and
+expose them in the `torch_np` module. This decorator is in charge of gathering
+all the inputs at runtime and normalizing them (i.e., converting `torch_np`
+objects to PyTorch counterparts) according to their annotations.
+
+We currently have four annotations (and small variations of them):
+- `ArrayLike`: The input can be a `torch_np.array`, a list of lists, a
+  scalar, or anything that NumPy would accept. It returns a `torch.Tensor`.
+- `DTypeLike`: Takes a `torch_np` dtype and returns a PyTorch dtype.
```
Comment (ev-br) [inline RFC.md:273]: Strictly speaking, it also accepts various multiple ways NumPy uses to label dtypes (strings, typecodes and who knows what, 'float32', 'f4' etc). Don't know if that's a level of detail we want to go here or not though.

File: RFC.md [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+functions form `torch_np.ndarray`s to `torch_np.ndarray`s, implementing the
+methods from the `ndarray` class is rather simple. We simply register all the
+free functions as methods or dunder methods appropriately. We also forward the
+properties to the properties within the PyTorch tensor and we are done.
+This creates a circular dependency which we break with a local import.
+
+### Testing
+
+The testing of the framework was done via ~~copying~~ vendoring tests from the
+NumPy test suit. Then, we would replace the NumPy imports with `torch_np`
```
Comment (ev-br) [inline RFC.md:305]:
```suggestion
NumPy test suite. Then, we would replace the NumPy imports with `torch_np`
```

File: RFC.md [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+imports. The failures on these tests were then triaged and discussed the
+priority of fixing each of them.
+
+In the (near) future, we plan to get some real world examples and run them
+through the library, to test its coverage and correctness.
```
Comment (ev-br) [inline RFC.md:310]: We did run it on several test examples, detailed in the README.

File: RFC.md [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+
+### Limitations
+
+A number of known limitations are tracked in the second part of the
+[OP of this issue](https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/73).
+There are some more in [this issue](https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/86).
```
Comment (ev-br) [inline RFC.md:316]: I think gh-86 has mainly run its course? 
Modulo maybe https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/86#issuecomment-1479524808, is it needed in the RFC

**Review by ev-br — APPROVED** [review]
This reads great. Several minor comments, all optional.

File: RFC.md (hunk 11 above)
Comment (rgommers) [reply RFC.md:16]: I'll try to fix that by upstreaming the relevant content here. Best for this RFC to be standalone.

**Comment by lezcano** [comment]
Addressed the last review. Once @rgommers does a last pass, we'll publish this one in the RFC repo.

File: RFC.md [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+are not being implemented? We could then assert that either that parameter has
+not been provided, and if it has, it has the same value as the default. The
+goal here would be to either use all the args of a function in its
+implementation, or mark explicitly those that we don't use.
+
+**Implmenting out**: In PyTorch, the `out` kwarg is, as the name says, a
+keyword-only argument. It is for this reason that, in PrimTorch, we were able
+to implement it as
+[a decorator](https://github.com/pytorch/pytorch/blob/ce4df4cc596aa10534ac6d54912f960238264dfd/torch/_prims_common/wrappers.py#L187-L282).
+This is not the case in NumPy. In NumPy `out` is a positional arg that is often
```
[the remark this answers is not in the document]
Comment (rgommers) [reply RFC.md:294]: Not sure what I meant two weeks ago, but this isn't true:
```python
>>> x = np.ones(3)
>>> np.sin(x, x)
array([0.84147098, 0.84147098, 0.84147098])
>>> x = np.ones(3)
>>> np.sin(x, out=x)
array([0.84147098, 0.84147098, 0.84147098])
```
I'll change this to "can be used both as a positional and a keyword argument".
